src/main/java/org/broadinstitute/hellbender/tools/walkers/vqsr/VariantRecalibrator.java

```diff
@@ -0,0 +1,852 @@
+        }
+
+        for( FeatureInput<VariantContext> variantFeature : resource ) {
+            dataManager.addTrainingSet( new TrainingSet( variantFeature ) );
+        }
+
+        if( !dataManager.checkHasTrainingSet() ) {
+            throw new UserException.CommandLineException(
+                    "No training set found! Please provide sets of known polymorphic loci marked with the training=true ROD binding tag. For example, -resource:hapmap,VCF,known=false,training=true,truth=true,prior=12.0 hapmapFile.vcf" );
+        }
+        if( !dataManager.checkHasTruthSet() ) {
+            throw new UserException.CommandLineException(
+                    "No truth set found! Please provide sets of known polymorphic loci marked with the truth=true ROD binding tag. For example, -resource:hapmap,VCF,known=false,training=true,truth=true,prior=12.0 hapmapFile.vcf" );
+        }
+
+        //TODO: this should be refactored/consolidated as part of
+        // https://github.com/broadinstitute/gatk/issues/121 and
+        // https://github.com/broadinstitute/gatk/issues/1116
+        // Initialize VCF header lines
+        final Set<VCFHeaderLine> hInfo = new HashSet<>();
+        hInfo.add(new VCFHeaderLine("source", this.getClass().getSimpleName()));
+        VariantRecalibrationUtils.addVQSRStandardHeaderLines(hInfo);
+        Set<VCFHeaderLine> actualLines = null;
+        SAMSequenceDictionary sequenceDictionary = null;
+        if (hasReference()) {
+            File refFile = referenceArguments.getReferenceFile();
+            sequenceDictionary= this.getReferenceDictionary();
+            actualLines = VcfUtils.updateHeaderContigLines(hInfo, refFile, sequenceDictionary, true);
+        }
+        else {
+            sequenceDictionary = getHeaderForVariants().getSequenceDictionary();
+            if (null != sequenceDictionary) {
+                actualLines = VcfUtils.updateHeaderContigLines(hInfo, null, sequenceDictionary, true);
+            }
+            else {
+                actualLines = hInfo;
+            }
+        }
+        //TODO: integrate lenient argument
+        recalWriter = GATKVariantContextUtils.createVCFWriter(new File(output), getBestAvailableSequenceDictionary(), true);
+        recalWriter.writeHeader( new VCFHeader(actualLines) );
+
+        for( int iii = 0; iii < REPLICATE * 2; iii++ ) {
+            replicate.add(Utils.getRandomGenerator().nextDouble());
+        }
+    }
+
+    //---------------------------------------------------------------------------------------------------------------
+    //
+    // map
+    //
+    //---------------------------------------------------------------------------------------------------------------
+
+    @Override
+    public void apply(VariantContext vc, ReadsContext readsContext, ReferenceContext ref, FeatureContext featureContext) {
+
+        // TODO: this variant came from the drivingvariants input; do we still need to query it again for overlapping ?
```

Comment:
My comment to `ApplyVQSR` applies here as well -- we should find a way to use the `vc` passed in to `apply()` directly, instead of doing a separate query against the driving input (which I think will hurt performance, and will incorrectly duplicate variants anyway if any share the same start position, unless I'm mistaken). Processing the vcs at a given start position one at a time instead of all at once should not really make a difference in the results -- if it does, we should understand exactly why. Worst case, we can accumulate variants in `apply()` until a new position is reached -- even that option is preferable to ignoring the variant passed in.